### PR TITLE
Add an icon button to open a file panel for actions with folder/file target

### PR DESCRIPTION
**Pull request by owenca**

### CODE REVIEW

**Comment by owenca** [comment]
I will work on changing the button to an icon button after this PR is merged.

File: sources/ActionView.cpp [hunk 1]
```diff
@@ -119,16 +171,25 @@ ActionView::GetAction() const
 }
 
 
+static void
+showOrHide(BControl* control, bool show)
```
Comment (humdingerb) [inline sources/ActionView.cpp]: Wouldn't naming the function just "ShowControl()" be OK?

Comment (humdingerb) [inline sources/ActionView.cpp]: Or maybe "SetVisibility()"?

Comment (owenca) [inline sources/ActionView.cpp]: I like `SetVisibility`.

**Comment by owenca** [comment]
I have changed the Open button to an icon button and updated the commit message.

File: sources/ActionView.cpp [hunk 2]
```diff
@@ -24,17 +28,38 @@ ActionView::ActionView(const char* name, BMessage* action, const int32& flags)
...
+		this, fValueBox->Bounds().Height());
 
 	BLayoutBuilder::Group<>(this, B_HORIZONTAL, B_USE_HALF_ITEM_SPACING)
 		.Add(fActionField, 0)
-		.Add(fValueBox)
+		.AddGroup(B_HORIZONTAL, 0)
+			.Add(fIconButton, 0)
+			.Add(fValueBox)
+			.End()
 		.Add(fAddRemoveButtons)
```
Comment (humdingerb) [inline sources/ActionView.cpp]: What do you think about adding a `.AddStrut(B_USE_HALF_ITEM_SPACING)` before the +/- buttons?
Separates the buttons a little bit that aren't part of the actual rule.
Would have to be added to the test view too.
Also, a bit wider window would give a little more space to the text control.

**Comment by humdingerb** [comment]
Looks very nice!
Type, Name, Size, and Location of the Test View could also do with that icon button... will that be next? :)

Q: how difficult is it to use different BRefFilters depending on the kind of test/action? Filer could limit the displayed files to e.g. zip-archives, or text/plain files in case of "Shell command".

File: sources/ActionView.cpp (hunk 2 above)
Comment (owenca) [inline sources/ActionView.cpp]: Good idea! Consider it done.

**Comment by owenca** [comment]
It makes sense to do the same for `TestView`. I will add it after this PR.

It shouldn't be hard to filter for file types, too. I think I can squeeze it in before this PR is merged. Is there a shell script file type instead of just text/plain?

**Comment by humdingerb** [comment]
> Is there a shell script file type instead of just text/plain?

Nope, those should all be just text/plain. OTOH, come to think of it, I'm not entirely sure how those BRefFilters deal with files on non-BFS partitions. On those, files are probably not identified by their MIME type...

Maybe it's worth an enhancement ticket for Haiku to deal gracefully with non-BFS volumes in this case.

**Comment by humdingerb** [comment]
Apparently, it's possible - and advisable - to not filter files with no filetype. See my quick conversation with DeadYak on IRC:

[16:26] <humdinger> hullo. files on non-bfs partitions don't normally show their (MIME) file type, right?
[16:27] <humdinger> how does a BFilePanel with a BRefFilter deal with that?
[16:27] <humdinger> should/could a BFilePanel in that case ignore the BRefsFilter?
[16:28] <DeadYak> humdinger: the same way a BRefFilter deals with a file on a BFS partition that doesn't yet have its mimetype set
[16:28] <DeadYak> that can occur either way
[16:28] <humdinger> so it's invisible?
[16:28] <DeadYak> a file on a BFS partition doesn't have that set until it's been identified, either via Tracker or mimeset -f or some other means
[16:29] <DeadYak> if you're filtering solely off mimetype? yes
[16:29] <DeadYak> either that or you have to allow unidentified files through as a fallback
[16:29] <DeadYak> for non-BFS filesystems it's typically the case that the filesystem driver has to supply synthetic types by some other means
[16:29] <humdinger> ah. that's possible?
[16:29] <DeadYak> i.e. the FAT driver does so based on file extensions
[16:29] <DeadYak> yes
[16:30] <humdinger> oh good!
[16:30] <DeadYak> either way, like I said that's a case you potentially come across on BFS too
[16:30] <DeadYak> so typically you want to fall back and let it through if you don't have a type to compare against

So, if you want to, you can look into that further.
We can also merge this PR and you can do such refinery later.

**Comment by owenca** [comment]
It's already supported, so it can be merged.

File: sources/TypedRefFilter.cpp [hunk 3]
```diff
@@ -60,11 +60,8 @@ TypedRefFilter::Filter(const entry_ref* ref, BNode* node, struct stat_beos* st,
 		return false;
 
 	// An empty file type means any file type
-	if (fFileType.Length() < 1)
+	if (fFileType.IsEmpty())
```
Comment (owenca) [inline sources/TypedRefFilter.cpp]: This is the fallback. :)

**Comment by humdingerb** [comment]
Merged. Thanks!
However, maybe look into mounting a e.g. Linux partition and check the file dialog of a "Shell command" action. It only shows folders there, even though there are plenty of text files around.

**Pull request merged**